Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp

```diff
@@ -304,6 +304,25 @@ bool WebLoaderStrategy::tryLoadingUsingPDFJSHandler(ResourceLoader& resourceLoad
 }
 #endif
 
+static RefPtr<DocumentLoader> policySourceDocumentLoaderForFrame(const LocalFrame& frame, bool isMainFrameNavigation = false)
+{
+    RefPtr mainFrame = dynamicDowncast<LocalFrame>(frame.mainFrame());
+    if (!mainFrame)
+        return { };
+
+    RefPtr mainFrameDocumentLoader = mainFrame->loader().policyDocumentLoader();
```

Comment:
Is this fallback implemented somewhere else? is it something worth sharing in WebCore?

Reply:
Indeed — we should be able to share some code between this, `WebLocalFrameLoaderClient::applyToDocumentLoader()`, and `WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction()`. Since this patch just moves the existing code out to a helper function though, I'll clean it up in a separate patch here: https://bugs.webkit.org/show_bug.cgi?id=257152. Since all the call sites are in WebKit2, I think I'll just add a helper function around `WebFrame.h` that we can use from all three places.

Thanks for the review!